### PR TITLE
Feature: Adding thread manager

### PR DESCRIPTION
**Pull request by eapg**
* Adding Thread Manager 

This manager will be implement to create, start, kill and know the status of a thread.

### CODE REVIEW

**Review by reynoldmorel — CHANGES_REQUESTED** [review]
This is better!

Now let's adapt it to our needs:

1. Rename `ThreadManager` to `AppEngineProcessor`.
2. Create an `AppEngineProcessConfig` that wiill contain any process config such as duration/timeout/interval, listeners like `on_init()`, `on_refresh()`, `before_shutdown()`, and whatever you want to add in order to setup your process. 
3. Create an `AppContext` that will contain all cross-shared information between processes
4. Create the `KitchenSimulatorProcess` and make the `AppEngineProcessor` register it.
5. Once you run `AppEngineProcessor.start()` it should start all registered processes (i.e `KitchenSimulatorProcess`).

**Comment by eapg** [comment]
* Added EngineAppProcessor
* Added EngineAppProcessHandler to handle process to be trigger when an specific event occur
* create 3 methods inside EngineAppProcesso add_subscribers(subscriber), remove_subscriber(subscriber) and star_threads() to trigger the AppEngineprocessHandler and start all the threads stored in the process handler.

File: src/tests/utils/fixtures/app_processor_config_fixture.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from src.core.engine.app_processor_config import AppProcessorConfig
+
+
+def build_app_engine_config(
+    id=None, interval=None, on_start=None, on_destroy=None, on_execute=None
+):
+    app_engine_config = AppProcessorConfig()
+    app_engine_config.id = id
```
Comment (reynoldmorel) [inline src/tests/utils/fixtures/app_processor_config_fixture.py:8]: Add a default value here

File: src/tests/core/engine/app_engine_processor_test.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from unittest.mock import patch
+
+from src.core.engine.app_engine_processor import AppEngineProcessor
+from src.core.engine.processors.kitchen_simulator import KitchenSimulator
+
+
+class TestAppEngineProcessor(unittest.TestCase):
+    @patch("src.core.engine.app_engine_processor.AppProcessorConfig")
+    @patch("src.core.engine.app_engine_processor.KitchenSimulator")
+    def test_mock_in_app_context(self, mock_kitchen, mock_config):
```
Comment (reynoldmorel) [inline src/tests/core/engine/app_engine_processor_test.py:11]: Make `mock_config` to return a `build_app_engine_config` when the instance is created. Remove the mock for KitchenSimulator.

Then assert that `len(app_engine_processor.app_context.processors) = 1` and `app_engine_processor.app_context.processors[0].app_processor_config.id` is equal to the id youi used here `build_app_engine_config`

File: src/tests/core/engine/app_engine_processor_test.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+        app_engine_processor = AppEngineProcessor()
+        self.assertEqual(*app_engine_processor.app_context.processors, mock_kitchen())
+        mock_config.assert_called_once()
+
+    @patch(
+        "src.core.engine.app_engine_processor.KitchenSimulator.start",
+        wraps=KitchenSimulator,
+    )
+    def test_start_in_context_run_once(self, mock_kitchen):
```
Comment (reynoldmorel) [inline src/tests/core/engine/app_engine_processor_test.py:21]: This test is not ok, you should mock `KitchenSimulator` and make it to return your own instance using a `build_kitchen_simulator_processor` function and then check if the start from that mock was called when you ran `app_engine_processor.start()`

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            id="test config", interval=10, on_start=None
+        )
+        app_engine_processor_context = build_app_engine_processor_context([])
+        test_class = ChildProcessor(app_engine_config, app_engine_processor_context)
+        test_thread_init.assert_called_once()
+
+    def test_app_engine_config(self):
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+        self.assertEqual(test_class.app_processor_config.id, app_engine_config.id)
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:38]: You can do this assertion in the tes above

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    def test_app_engine_processor_context(self):
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:40]: Same here

File: src/tests/core/engine/app_engine_processor_integration_test.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+from src.core.engine.processors.kitchen_simulator import KitchenSimulator
+
+
+class AppEngineProcessorIntegrationTest(TestCase):
+
+    @patch(
+        "src.core.engine.app_engine_processor.KitchenSimulator.process",
+        wraps=KitchenSimulator,
+    )
+    def test_process_run(self, mock_process):
```
Comment (reynoldmorel) [inline src/tests/core/engine/app_engine_processor_integration_test.py:13]: Rename `mock_process`  to `mock_kitchen_simulator_process`

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        app_engine_processor_context = build_app_engine_processor_context([])
+        test_class = ChildProcessor(None, app_engine_processor_context)
+        self.assertEqual(
+            test_class.app_context.processors, app_engine_processor_context.processors
+        )
+
+    def test_self_pause_is_false(self):
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:47]: Same here

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        test_class = ChildProcessor()
+        self.assertEqual(test_class.pause, False)
+
+    def test_self_destroy_is_false(self):
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:51]: same for this one

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        test_class = ChildProcessor()
+        self.assertEqual(test_class.destroy, False)
+
+    def test_last_execution_time(self):
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+        test_class.start()
+        self.assertAlmostEqual(test_class.last_execution_time, datetime.now())
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:59]: Mock `datetime.now()` and use your own fixed number.

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+    def test_process_method_was_call_with(self):
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+
+        with patch.object(ChildProcessor, "process") as mock:
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:65]: Could you follow same pattern as before?

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    def test_on_destroy_when_destroy_set_true(self):
+        mock_function = Mock(wraps=lambda: print("on destroy"))
+        app_engine_config = build_app_engine_config(
+            id="test config", interval=10, on_destroy=mock_function
+        )
+        test_class = ChildProcessor(app_engine_config)
+        test_class.start()
+        self.assertTrue(test_class.is_alive())
+        test_class.destroy = True
+        time.sleep(15)
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:105]: hmm this is tricky, could you mock `time.sleep` implementation and control how much time sleep function will last? It is better for scalability and flexibility

File: src/tests/core/engine/processors/abstract_processor_test.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        self.assertFalse(test_class.is_alive())
+
+    def test_set_pause_true(self):
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+
+        with patch.object(ChildProcessor, "process") as mock:
```
Comment (reynoldmorel) [inline src/tests/core/engine/processors/abstract_processor_test.py:112]: Use annotation as previously, or use this for every test

**Review by reynoldmorel — CHANGES_REQUESTED** [review]
Rename every `mock_*` variable to `mocked_*`.

This a nice first try to these kinf of tests.

Thanks

File: src/tests/utils/fixtures/app_processor_config_fixture.py (hunk 1 above)
Comment (eapg) [reply src/tests/utils/fixtures/app_processor_config_fixture.py:8]: Done

File: src/tests/core/engine/app_engine_processor_integration_test.py (hunk 6 above)
Comment (eapg) [reply src/tests/core/engine/app_engine_processor_integration_test.py:13]: Done. Renamed to mocked_kitchen_simulator_process

File: src/tests/core/engine/app_engine_processor_test.py (hunk 2 above)
Comment (eapg) [reply src/tests/core/engine/app_engine_processor_test.py:11]: Done.

File: src/tests/core/engine/processors/abstract_processor_test.py (hunk 4 above)
Comment (eapg) [reply src/tests/core/engine/processors/abstract_processor_test.py:38]: done

File: src/tests/core/engine/processors/abstract_processor_test.py (hunk 5 above)
Comment (eapg) [reply src/tests/core/engine/processors/abstract_processor_test.py:40]: Done.

File: src/tests/core/engine/processors/abstract_processor_test.py (hunk 7 above)
Comment (eapg) [reply src/tests/core/engine/processors/abstract_processor_test.py:47]: Done.

File: src/tests/core/engine/processors/abstract_processor_test.py (hunk 8 above)
Comment (eapg) [reply src/tests/core/engine/processors/abstract_processor_test.py:51]: Done.

File: src/tests/core/engine/processors/abstract_processor_test.py (hunk 10 above)
Comment (eapg) [reply src/tests/core/engine/processors/abstract_processor_test.py:65]: Done.

File: src/tests/core/engine/processors/abstract_processor_test.py (hunk 12 above)
Comment (eapg) [reply src/tests/core/engine/processors/abstract_processor_test.py:112]: Done.

**Pull request merged**